### PR TITLE
Move section index to property

### PR DESCRIPTION
**Pull request by amonshiz**
## Changes in this pull request
This pull request removes the `sectionForSectionController:` method from the `IGListCollectionContext` protocol so that the protocol is exclusively for presentation methods.

This should not add new functionality, but rather makes the index directly accessible on the section controllers themselves. This change makes sense because at no time will there be an update to the list that the list adapter is unaware of and so it will always be able to set and update any indexes for a section controller that has changed.

Issue fixed: #609

### Checklist

- [X] All tests pass. Demo project builds and runs.
- [X] I added tests, an experiment, or detailed why my change isn't tested.
- [X] I added an entry to the `CHANGELOG.md` for any breaking changes, enhancements, or bug fixes.
- [X] I have reviewed the [contributing guide](https://github.com/Instagram/IGListKit/blob/master/.github/CONTRIBUTING.md)

### CODE REVIEW

File: Tests/IGListAdapterE2ETests.m [hunk 1]
```diff
@@ -1477,25 +1477,25 @@ - (void)test_whenInsertingItemsTwice_withDataUpdatedTwice_thatAllUpdatesAppliedW
...
+//        object.value = @2;
+//        [batchContext deleteInSectionController:sectionController atIndexes:[NSIndexSet indexSetWithIndex:0]];
+//        [batchContext deleteInSectionController:sectionController atIndexes:[NSIndexSet indexSetWithIndex:0]];
+//    } completion:^(BOOL finished2) {
+//        XCTAssertEqual([self.collectionView numberOfSections], 1);
+//        XCTAssertEqual([self.collectionView numberOfItemsInSection:0], 2);
+//        [expectation fulfill];
+//    }];
+//    [self waitForExpectationsWithTimeout:15 handler:nil];
+//}
```
Comment (jessesquires) [inline Tests/IGListAdapterE2ETests.m]: 🤔

File: Tests/IGListStackSectionControllerTests.m [hunk 2]
```diff
@@ -246,12 +246,12 @@ - (void)test_whenQueryingSectionControllerSection_thatSectionMatchesStackSection
...
-    XCTAssertEqual([section11.collectionContext sectionForSectionController:section11], 0);
-    XCTAssertEqual([section12.collectionContext sectionForSectionController:section12], 0);
-    XCTAssertEqual([section21.collectionContext sectionForSectionController:section21], 1);
-    XCTAssertEqual([section22.collectionContext sectionForSectionController:section22], 1);
+    XCTAssertEqual(stack1.sectionIndex, 0);
+    XCTAssertEqual(stack2.sectionIndex, 1);
+    XCTAssertEqual(section11.sectionIndex, 0);
+    XCTAssertEqual(section12.sectionIndex, 0);
+    XCTAssertEqual(section21.sectionIndex, 1);
+    XCTAssertEqual(section22.sectionIndex, 1);
```
Comment (jessesquires) [inline Tests/IGListStackSectionControllerTests.m]: add more complex cases here?

File: Tests/IGListAdapterTests.m [hunk 3]
```diff
@@ -1133,4 +1133,32 @@ - (void)test_whenQueryingInsetContainerSize_thatResultIsBoundsInsetByContent {
...
+
+    IGListSectionController<IGListSectionType> *controller1b = [self.adapter sectionControllerForObject:@1];
+    XCTAssertEqual(controller1b.sectionIndex, 1);
+    XCTAssertFalse(controller1b.isFirstSection);
+
+    XCTAssertEqual([self.adapter sectionControllerForObject:@2].sectionIndex, 2);
+    XCTAssertEqual([self.adapter sectionControllerForObject:@3].sectionIndex, 3);
+    XCTAssertTrue([self.adapter sectionControllerForObject:@3].isLastSection);
+}
+
```
Comment (jessesquires) [inline Tests/IGListAdapterTests.m]: we should probably test: (1) append, (2) insert in middle

**Comment by jessesquires** [comment]
`sectionIndex` will always be correct, right? even after ListAdapter mutations?

**Comment by jessesquires** [comment]
let's make sure we handle the cases that @Adlai-Holler mentioned in the issue

**Comment by amonshiz** [comment]
All the existing tests pass, so at least we have those mutations covered. Would it be worth adding a list adapter mutations specific testing file?

This one test was to cover the case mentioned in the task. However, speaking with Ryan it seems this *shouldn't* be passing currently so I will need to debug a bit.

File: Tests/IGListStackSectionControllerTests.m (hunk 2 above)
Comment (amonshiz) [inline Tests/IGListStackSectionControllerTests.m]: My preference would just be to 86 the stack controller ...

File: Tests/IGListAdapterE2ETests.m (hunk 1 above)
Comment (amonshiz) [inline Tests/IGListAdapterE2ETests.m]: science

File: Source/Internal/IGListSectionMap.h [hunk 4]
```diff
@@ -38,7 +38,7 @@ IGLK_SUBCLASSING_RESTRICTED
  @param objects The objects in the collection.
  @param sectionControllers The section controllers that map to each object.
  */
-- (void)updateWithObjects:(NSArray <id <NSObject>> *)objects sectionControllers:(NSArray <id <NSObject>> *)sectionControllers;
+- (void)updateWithObjects:(NSArray <id <NSObject>> *)objects sectionControllers:(NSArray <IGListSectionController *> *)sectionControllers;
```
Comment (amonshiz) [inline Source/Internal/IGListSectionMap.h]: at some point this method was not updated to the `IGListSectionController<IGListSectionType> *` type and was missed in a few refactorings. the next method down, `sectionControllerForSection:` returned an `IGListSectionController *` so the types were a significant mismatch and needed to be updated.

File: Tests/IGListSectionMapTests.m [hunk 5]
```diff
@@ -23,39 +23,39 @@ @implementation IGListSectionMapTests
 
 - (void)test_whenUpdatingItems_thatArraysAreEqual {
     NSArray *objects = @[@0, @1, @2];
-    NSArray *sectionControllers = @[@"a", @"b", @"c"];
+    NSArray *sectionControllers = @[[IGListTestSection new], [IGListTestSection new], [IGListTestSection new]];
```
Comment (amonshiz) [inline Tests/IGListSectionMapTests.m]: all of these are just so that the test actually respects the types that `IGListSectionMap` require.

**Comment by jessesquires** [comment]
@amonshiz -- looks like this is good to go, except for resolving conflicts?

**Comment by amonshiz** [comment]
will do!

**Pull request closed without merging**